src/multiversioning.zig

```diff
@@ -0,0 +1,1372 @@
+    assert(name.len < test_elf_name_length_max);
+    assert(index < elf_header.e_shnum);
+
+    const offset: u64 = elf_header.e_shoff + @as(u64, @sizeOf(elf.Elf64_Shdr)) * index;
+    const elf_section_header = std.mem.bytesAsValue(
+        elf.Elf64_Shdr,
+        buffer[offset..][0..@sizeOf(elf.Elf64_Shdr)],
+    );
+    elf_section_header.sh_name = test_elf_name_length_max * index;
+    try expect_any_error(parse_elf(buffer));
+
+    stdx.copy_disjoint(.inexact, u8, string_table[elf_section_header.sh_name..], name);
+    try expect_any_error(parse_elf(buffer));
+    string_table[elf_section_header.sh_name..][name.len] = 0;
+    try expect_any_error(parse_elf(buffer));
+    elf_section_header.sh_offset = 8192 * index;
+    try expect_any_error(parse_elf(buffer));
+
+    return elf_section_header;
+}
+
+// Not quite a fuzzer, but build up an ELF, checking that there's an error after each step, with a
+// full range of values is the undefined intermediate bits.
+test "parse_elf" {
+    var buffer: [32768]u8 align(8) = undefined;
+    for (0..256) |i| {
+        @memset(&buffer, @as(u8, @intCast(i)));
+
+        const elf_header = try test_elf_build_header(&buffer);
+        const string_table = try test_elf_build_string_table(&buffer, elf_header);
+
+        // The string table can't be 0, and the .tb_mvb and .tb_mvh sections need to be the second
+        // last and last sections in the file respectively. Pad the 0 index with a no-op section.
+        _ = try test_elf_build_section(&buffer, string_table, elf_header, 0, ".tb_nop");
+
+        const section_mvb = try test_elf_build_section(
+            &buffer,
+            string_table,
+            elf_header,
+            2,
+            ".tb_mvb",
+        );
+        // So it overlaps on purpose, to check the MultiversionBodyOverlapsHeader assert.
+        section_mvb.sh_size = 16384;
+
+        const section_mvh = try test_elf_build_section(
+            &buffer,
+            string_table,
+            elf_header,
+            3,
+            ".tb_mvh",
+        );
+        section_mvh.sh_size = 8192; // @sizeOf(MultiversionHeader), but hardcoded.
+
+        try std.testing.expectError(error.MultiversionBodyOverlapsHeader, parse_elf(&buffer));
+
+        section_mvb.sh_size = 8192;
+        const parsed = try parse_elf(&buffer);
+
+        assert(parsed.body == 16384);
+        assert(parsed.header == 24576);
+    }
+}
+
+pub fn validate(allocator: std.mem.Allocator, exe_path: [:0]const u8) !void {
+    if (builtin.target.os.tag != .linux) @panic("only linux is supported for validate()");
+
+    var io = try IO.init(32, 0);
+    defer io.deinit();
+
+    const absolute_exe_path = try allocator.dupeZ(
+        u8,
+        try std.fs.cwd().realpathAlloc(allocator, exe_path),
+    );
+
+    var multiversion = try Multiversion.init(
+        allocator,
+        &io,
+        absolute_exe_path,
+    );
+    defer multiversion.deinit(allocator);
+
+    multiversion.start();
```

Comment:
Setting the error here is only for event loop errors, which are rare. We expect most of them to be set on the stage through `handle_error()`. But, in the new version I've made it return `self.stage.err` if there is one, since that's a nice pattern for the utility functions like `print_information` and co.